emma-language/src/main/scala/org/emmalanguage/ast/AST.scala

```diff
@@ -90,6 +90,31 @@ trait AST extends CommonAST
     case api.TypeQuote(tpe) => api.Type.tree(tpe)
   }.andThen(_.tree)
 
+  /**
+   * Rewrites `A.this.x` as `x` if `A` is shadowed another symbol in the owner chain.
+   *
+   * Can be used before `showCode` to avoid printing invalid trees such as
+   *
+   * {{{
+   * class A {
+   *   val x = 42;
+   *   class A {
+   *     scala.Predef.println(A.this.x)
+   *   }
+   * }
+   * }}}
+   */
+  lazy val removeShadowedThis =
+    if (shadowedOwners.isEmpty) identity[u.Tree] _
+    else api.TopDown.transform {
+      case api.Sel(api.This(encl), api.Id(member))
```

Comment:
This will never match. `Sel` returns a pair of qualifier (`u.Tree`) and member (`u.Symbol`).